src/simulators/circuit_executor.hpp

```diff
@@ -0,0 +1,1174 @@
+      parallel_shots_ = 1;
+      parallel_state_update_ =
+          std::max<int>({1, max_parallel_threads_ / parallel_experiments_});
+      return;
+    }
+    break;
+  }
+  case Method::density_matrix:
+  case Method::superop:
+  case Method::tensor_network: {
+    if (circ.shots == 1 || num_process_per_experiment_ > 1 ||
+        check_measure_sampling_opt(circ)) {
+      parallel_shots_ = 1;
+      parallel_state_update_ =
+          std::max<int>({1, max_parallel_threads_ / parallel_experiments_});
+      return;
+    }
+    break;
+  }
+  case Method::extended_stabilizer:
+    break;
+  default:
+    throw std::invalid_argument(
+        "Cannot set parallelization for unresolved method.");
+  }
+
+  // Use a local variable to not override stored maximum based
+  // on currently executed circuits
+  const auto max_shots =
+      (max_parallel_shots_ > 0)
+          ? std::min({max_parallel_shots_, max_parallel_threads_})
+          : max_parallel_threads_;
+
+  // If we are executing circuits in parallel we disable
+  // parallel shots
+  if (max_shots == 1 || parallel_experiments_ > 1) {
+    parallel_shots_ = 1;
+  } else {
+    // Parallel shots is > 1
+    // Limit parallel shots by available memory and number of shots
+    // And assign the remaining threads to state update
+    int circ_memory_mb =
+        required_memory_mb(circ, noise) / num_process_per_experiment_;
+    size_t mem_size =
+        (sim_device_ == Device::GPU) ? max_gpu_memory_mb_ : max_memory_mb_;
+    if (mem_size < circ_memory_mb)
+      throw std::runtime_error(
+          "a circuit requires more memory than max_memory_mb.");
+    // If circ memory is 0, set it to 1 so that we don't divide by zero
+    circ_memory_mb = std::max<int>({1, circ_memory_mb});
+
+    int shots = circ.shots;
+    parallel_shots_ = std::min<int>(
+        {static_cast<int>(mem_size / (circ_memory_mb * 2)), max_shots, shots});
+  }
+  parallel_state_update_ =
+      (parallel_shots_ > 1)
+          ? std::max<int>({1, max_parallel_threads_ / parallel_shots_})
+          : std::max<int>({1, max_parallel_threads_ / parallel_experiments_});
+}
+
+template <class state_t>
+void Executor<state_t>::run_circuit(Circuit &circ,
+                                    const Noise::NoiseModel &noise,
+                                    const Config &config, const Method method,
+                                    const Device device,
+                                    ExperimentResult &result) {
+  // Start individual circuit timer
+  auto timer_start = myclock_t::now(); // state circuit timer
+
+  // Execute in try block so we can catch errors and return the error message
+  // for individual circuit failures.
+  try {
+    // set configuration
+    method_ = method;
+    sim_device_ = device;
+
+    set_config(config);
+    set_parallelization(circ, noise);
+
+    // Rng engine (this one is used to add noise on circuit)
+    RngEngine rng;
+    rng.set_seed(circ.seed);
```

Comment:
`rng.set_seed(circ.seed);` is called in later (`run_circuit_with_sampling` or `run_circuit_shots `). It is better to use different seeds for them.

Reply:
Now initial rng used for noise sampling is reused in `run_circuit_xxx` functions.
I think this is the strategy used in the older code.